planner/core/task.go

```diff
@@ -557,28 +557,186 @@ func (p *PhysicalHashJoin) attach2Task(tasks ...task) task {
 	return task
 }
 
+// for different decimal scale and precision, tiflash uses different underlying type.
+// Here we check the scale and precision to decide whether conversion is a must.
+func needDecimalConvert(tp *types.FieldType, dec int, l int) bool {
+	if tp.Decimal != dec {
+		return true
+	}
+	if tp.Flen >= 0 && tp.Flen <= 9 && l >= 0 && l <= 9 {
+		return false
+	}
+	if tp.Flen > 9 && tp.Flen <= 18 && l > 9 && l <= 18 {
+		return false
+	}
+	if tp.Flen > 18 && tp.Flen <= 38 && l > 18 && l <= 38 {
+		return false
+	}
+	if tp.Flen > 38 && tp.Flen <= 65 && l > 38 && l <= 65 {
+		return false
+	}
+	return true
+}
+
+func negotiateCommonType(lType, rType *types.FieldType) (*types.FieldType, bool, bool) {
+	lExtend := 0
+	rExtend := 0
+	cDec := rType.Decimal
+	if lType.Decimal < rType.Decimal {
+		lExtend = rType.Decimal - lType.Decimal
+	} else if lType.Decimal > rType.Decimal {
+		rExtend = lType.Decimal - rType.Decimal
+		cDec = lType.Decimal
+	}
+	lLen, rLen := lType.Flen+lExtend, rType.Flen+rExtend
+	cLen := lLen
+	if rLen > cLen {
+		cLen = rLen
+	}
+	if cLen > 65 {
+		cLen = 65
+	}
+	cType := types.NewFieldType(mysql.TypeNewDecimal)
+	cType.Decimal = cDec
+	cType.Flen = cLen
```

Comment:
should other fields be considered? such as flag...

Reply:
fixed